File: scripts/Unique Paths/Unique Paths.java
class Solution {
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];
        
        for(int i = 0; i < m; i ++) {
            for(int j = 0; j < n; j ++) {
                dp[i][j] = -1;
            }
        }
        
        return helper(m, 0, n, 0, dp);
    }
    
    private int helper(int m, int i, int n, int j, int[][] dp) {
        if(i == m || j == n) {
            return 0;
        }
        
        if(i == m-1 && j == n-1) {
            dp[i][j] = 1;
        }
        
        if(dp[i][j] == -1) {
            dp[i][j] = helper(m, i+1, n, j, dp) + helper(m, i, n, j+1, dp);
        }
        
        return dp[i][j];
    }
}

File: scripts/algorithms/U/Unique Paths/Unique Paths.py
// Runtime: 39 ms (Top 53.95%) | Memory: 17.30 MB (Top 9.72%)

class Solution:
    def uniquePaths(self, m, n):
        dp = [[1]*n for i in range(m)]
        for i, j in product(range(1, m), range(1, n)):
            dp[i][j] = dp[i-1][j] + dp[i][j-1]
        return dp[-1][-1]
